### PR TITLE
Create Map comparison utilities and convert Hooks to Typescript

### PR DESCRIPTION
**Pull request by t3chguy**
Subset of https://github.com/matrix-org/matrix-react-sdk/pull/4745

### CODE REVIEW

File: src/components/views/rooms/RoomList.tsx [hunk 1]
```diff
@@ -245,6 +245,7 @@ export default class RoomList extends React.PureComponent<IProps, IState> {
         if (doUpdate) {
             // We have to break our reference to the room list store if we want to be able to
             // diff the object for changes, so do that.
+            // @ts-ignore - ITagMap is ts-ignored so this will have to be too
```
Comment (JorikSchellekens) [inline src/components/views/rooms/RoomList.tsx:248]: Not looking to change it but :'(

File: src/utils/maps.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+import { arrayDiff, arrayMerge, arrayUnion } from "./arrays";
+
+/**
+ * Determines the keys added, changed, and removed between two Maps.
+ * For changes, simple triple equal comparisons are done, not in-depth tree checking.
+ * @param a The first Map. Must be defined.
+ * @param b The second Map. Must be defined.
+ * @returns The difference between the keys of each Map.
+ */
+export function mapDiff<K, V>(a: Map<K, V>, b: Map<K, V>): { changed: K[], added: K[], removed: K[] } {
```
Comment (JorikSchellekens) [inline src/utils/maps.ts:26]: It doesn't look like anything is using these?

Comment (t3chguy) [reply src/utils/maps.ts:26]: Correct you are, https://github.com/matrix-org/matrix-react-sdk/pull/4745 does and just plucked out the reusable stuff to shrink the diff of that mammoth

Comment (JorikSchellekens) [reply src/utils/maps.ts:26]: Damn, that is hefty

**Pull request merged**